GRUPO 14. IEBS.
MASTER EN BIG DATA Y DATA SCIENCE

#INTRODUCCIÓN

CARGAMOS LAS LIBRERÍAS A OCUPAR

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
#Montamos Drive con las bases de datos
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Cargamos la base del 2019, y luego concatenamos las de los otros años.
base1 = pd.read_excel('/content/drive/MyDrive/GLOBAL_PROJECT_G14/BASES_FINALES/FL_2019.xlsx', sheet_name='Hoja1')
#Repetimos el procedimiento
for i in range(2020,2025):
    base = pd.read_excel(f'/content/drive/MyDrive/GLOBAL_PROJECT_G14/BASES_FINALES/FL_{i}.xlsx', sheet_name='Hoja1')
    base1 = pd.concat([base1,base])

### **Inspección de datos**

In [ ]:
base1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 689169 entries, 0 to 58873
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   S_PERIODO         689169 non-null  int64         
 1   S_MAT             689169 non-null  int64         
 2   S_GENERO          689169 non-null  object        
 3   D_FEC_NACI        689169 non-null  datetime64[ns]
 4   N_EDAD            689169 non-null  int64         
 5   S_GRP_GENERACION  689169 non-null  object        
 6   S_EST_CIVIL       68784 non-null   object        
 7   S_TIP_CTRTO       20127 non-null   object        
 8   D_FEC_CTRTO       689169 non-null  datetime64[ns]
 9   S_JRND_LAB        10181 non-null   object        
 10  S_GRD_SALARIAL    10066 non-null   object        
 11  S_PUESTO          689169 non-null  object        
 12  D_FEC_INI_PSTO    689169 non-null  datetime64[ns]
 13  N_TPO_PSTO        689169 non-null  int64         
 14  S_DIVISION

Hay columnas con nulos.

In [ ]:
base1.head(10)

S_PERIODO  S_MAT S_GENERO D_FEC_NACI  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
0     201901  16537        F 1980-09-27      38     GENERACION X         NaN   
1     201901  17009        F 1985-03-27      33      MILLENNIALS         NaN   
2     201901  15635        M 1983-11-13      35      MILLENNIALS         NaN   
3     201901  11793        M 1983-06-18      35      MILLENNIALS         NaN   
4     201901  18596        F 1981-11-15      37      MILLENNIALS         NaN   
5     201901  12191        M 1977-01-18      41     GENERACION X         NaN   
6     201901  17764        F 1985-09-23      33      MILLENNIALS         NaN   
7     201901  13839        F 1986-02-21      32      MILLENNIALS         NaN   
8     201901  15242        M 1986-10-24      32      MILLENNIALS         NaN   
9     201901  78177        M 1988-11-06      30      MILLENNIALS         NaN   

  S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
0         NaN  2013-02-01        NaN            NaN   
1         NaN  2013-05-02        NaN            NaN   
2         NaN  2012-08-15        NaN            NaN   
3         NaN  2009-05-06        NaN            NaN   
4         NaN  2014-02-03        NaN            NaN   
5         NaN  2010-02-01        NaN            NaN   
6         NaN  2013-09-16        NaN            NaN   
7         NaN  2011-09-15        NaN            NaN   
8         NaN  2012-06-22        NaN            NaN   
9         NaN  2017-02-01        NaN            NaN   

                                  S_PUESTO D_FEC_INI_PSTO  N_TPO_PSTO  \
0        EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
1                     ASESOR DE NEGOCIOS 2     2018-04-01           9   
2                     ASESOR DE NEGOCIOS 1     2018-10-25           3   
3  ANALISTA SENIOR DE POLITICAS Y PROCESOS     2018-08-29           5   
4                     ASESOR DE NEGOCIOS 3     2014-08-15          53   
5          SUPERVISOR DE BANCA DE SERVICIO     2016-11-01          26   
6                     ASESOR DE NEGOCIOS 1     2018-12-15           1   
7                    EJECUTIVO DE NEGOCIOS     2017-10-21          15   
8             JEFE DE BANCA DE SERVICIO II     2015-04-01          45   
9                     ASESOR DE NEGOCIOS 2     2018-10-25           3   

                         S_DIVISION S_AGENCIA S_RED_STAFF  
0  DIVISION DE SOPORTE CENTRALIZADO       NaN         NaN  
1                          NEGOCIOS       NaN         NaN  
2                          NEGOCIOS       NaN         NaN  
3                           RIESGOS       NaN         NaN  
4                          NEGOCIOS       NaN         NaN  
5  DIVISION DE SOPORTE CENTRALIZADO       NaN         NaN  
6                          NEGOCIOS       NaN         NaN  
7                          NEGOCIOS       NaN         NaN  
8  DIVISION DE SOPORTE CENTRALIZADO       NaN         NaN  
9                          NEGOCIOS       NaN         NaN

In [ ]:
base1['S_MAT'].value_counts() #Hay empleados que aparecen todos los meses. Desde 2019 hasta Junio 2024

S_MAT
16537    66
17652    66
60228    66
81416    66
77341    66
         ..
93618     1
93639     1
82401     1
93845     1
97568     1
Name: count, Length: 23508, dtype: int64

In [ ]:
#Comenzamos a revisar los puestos de tecnología
base1['S_DIVISION'].value_counts()

S_DIVISION
NEGOCIOS                                                               463877
DIVISION DE SOPORTE CENTRALIZADO                                       101480
RIESGOS                                                                 47665
TECNOLOGIA Y OPERACIONES                                                32101
GESTION DE DESARROLLO HUMANO                                             8720
DIVISION DE NEGOCIOS                                                     8359
FINANZAS                                                                 4686
AUDITORIA                                                                3219
G.G ADJ. DE INTERNACIONALIZACION E INCLUSION FINANCIERA                  3088
TRIBU - SEGMENTOS Y PRODUCTOS                                            2561
TRANSFORMACION                                                           2447
DIVISION DE MARKETING                                                    2437
TRIBU - CANAL TRADICIONAL                                                2028
LEGAL                                                                    1227
TECNOLOGIA                                                                681
GERENCIA GENERAL ADJUNTA DE TRANSFORMACION Y GESTION ORGANIZACIONAL       620
TRIBU - CANALES DIGITALES Y SOCIOS                                        537
GERENCIA GENERAL ADJUNTA DE GESTION DE NEGOCIOS                           534
CUMPLIMIENTO Y ETICA                                                      466
DIVISION DE INTERNACIONALIZACION                                          410
GERENCIA GENERAL ADJUNTA DE GESTION ORGANIZACIONAL                        335
GERENCIA GENERAL                                                          304
DIVISION DE ADMINISTRACION                                                299
PLANEAMIENTO ESTRATEGICO                                                  194
GESTION ESTRATEGICA Y SERVICIOS PARA LOS CLIENTES                         156
GESTION ESTRATEGICA Y SERV PARA LOS CLIENTES                              153
DIVISION DE GDH                                                           129
DIVISION DE FINANZAS                                                       68
TRIBU CANAL TRADICIONAL                                                    61
TRIBU SEGMENTOS Y PRODUCTOS                                                49
DIVISION DE TRANSFORMACION                                                 48
GERENCIA GENERAL ADJUNTA                                                   37
DIVISION LEGAL                                                             19
TRIBU CANALES DIGITALES Y SOCIOS                                           15
DESTACADOS AL EXTRANJERO                                                   13
TRIBU ECOSISTEMAS                                                          10
GERENCIA DE PROYECTOS ESTRATEGICOS                                          7
GERENCIA DE INTERNACIONALIZACION E INCLUSION FINANCIERA                     1
GERENCIA GENERAL DE GESTION ORGANIZACIONAL                                  1
Name: count, dtype: int64

In [ ]:
del(base) #Para no ocupar tanto espacio en memoria

In [ ]:
for puesto in base1['S_PUESTO'].unique():
  if 'DAT' in puesto: print(puesto)                                      # Nos quedamos con las divisiones que contienen puestos relacionados a DATOS.

SUPERVISOR DE BASE DE DATOS
ANALISTA SENIOR DE BASE DE DATOS
ESPECIALISTA DE BASE DE DATOS
ANALISTA SENIOR DE SEGURIDAD DE BASE DE DATOS
GERENTE DE SERVICIO DE TRIBU DE DATOS
ESPECIALISTA DE GOBIERNO DE DATOS
ANALISTA SENIOR DE INGENIERIA DE DATOS
ESPECIALISTA DE CONSULTORIA DE DATOS
SUPERVISOR DE INGENIERIA DE DATOS
SUBGERENTE DE AUDITORIA CONTINUA Y ANALITICA DE DATOS
ANALISTA SENIOR DE GOBIERNO DE DATOS
AUDITOR JR DE ANALISIS DE DATOS Y MODELOS
PRODUCT OWNER DE PLANEAMIENTO Y DATA
ESPECIALISTA DE INGENIERIA DE DATOS
ANALISTA SENIOR DE DATA Y DESPLIEGUE DE PRICING
SUBGERENTE DE CONSULTORIA DE DATOS
ESPECIALISTA DE PLANEAMIENTO Y DATA
ARQUITECTO DE DATOS
ESPECIALISTA DE ARQUITECTURA DE DATOS
SUBGERENTE DE ARQUITECTURA DE DATOS
AUDITOR DE ANALISIS DE DATOS Y MODELOS
GERENTE DE GOBIERNO DE DATOS
ANALISTA DE DATOS DE DESARROLLO DEL EMPRENDEDOR
ASISTENTE DE SEGURIDAD DE BASE DE DATOS
ESPECIALISTA DE ENRIQUECIMIENTO DE DATOS CRM
GERENTE DE INGENIERIA DE DATOS
SUBGERENTE DE INGENIERIA DE DA

In [ ]:
#Checamos el área de los puestos que contienen "DATOS" en su nombre.
puesto_datos = base1[base1['S_PUESTO'].str.contains('DAT')]
puesto_datos['S_DIVISION'].unique()

array(['DIVISION DE SOPORTE CENTRALIZADO',
       'GERENCIA GENERAL ADJUNTA DE GESTION ORGANIZACIONAL',
       'G.G ADJ. DE INTERNACIONALIZACION E INCLUSION FINANCIERA',
       'TRANSFORMACION', 'AUDITORIA', 'TRIBU - SEGMENTOS Y PRODUCTOS',
       'FINANZAS', 'TECNOLOGIA Y OPERACIONES', 'RIESGOS', 'TECNOLOGIA',
       'TRIBU SEGMENTOS Y PRODUCTOS', 'DIVISION DE FINANZAS'],
      dtype=object)

Vemos que la división "TRANSFORMACIÓN" ya está incluída en estas divisiones, por lo que nos quedaremos también con los puestos de SCRUM y de AGILE, sin quedarnos con toda la división donde se encuentran.

### Descripción de empleados.

La columna "S_RED_STAFF" contiene información de los empleados y a que agencia pertenecen. Sin embargo, si son parte de "la red", son empleados de call centers, por lo que, se van a eliminar dichos empleados.

In [ ]:
base1['S_RED_STAFF'].value_counts()

S_RED_STAFF
RED      59184
STAFF     9600
Name: count, dtype: int64

In [ ]:
base1_filtrado = base1[base1['S_RED_STAFF'] != 'RED']
#Eliminamos la columna agencia, pues solo contiene información de empleados de la red.
base1_filtrado = base1_filtrado.drop(columns=['S_AGENCIA'])
base1_filtrado.head(10)

S_PERIODO  S_MAT S_GENERO D_FEC_NACI  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
0     201901  16537        F 1980-09-27      38     GENERACION X         NaN   
1     201901  17009        F 1985-03-27      33      MILLENNIALS         NaN   
2     201901  15635        M 1983-11-13      35      MILLENNIALS         NaN   
3     201901  11793        M 1983-06-18      35      MILLENNIALS         NaN   
4     201901  18596        F 1981-11-15      37      MILLENNIALS         NaN   
5     201901  12191        M 1977-01-18      41     GENERACION X         NaN   
6     201901  17764        F 1985-09-23      33      MILLENNIALS         NaN   
7     201901  13839        F 1986-02-21      32      MILLENNIALS         NaN   
8     201901  15242        M 1986-10-24      32      MILLENNIALS         NaN   
9     201901  78177        M 1988-11-06      30      MILLENNIALS         NaN   

  S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
0         NaN  2013-02-01        NaN            NaN   
1         NaN  2013-05-02        NaN            NaN   
2         NaN  2012-08-15        NaN            NaN   
3         NaN  2009-05-06        NaN            NaN   
4         NaN  2014-02-03        NaN            NaN   
5         NaN  2010-02-01        NaN            NaN   
6         NaN  2013-09-16        NaN            NaN   
7         NaN  2011-09-15        NaN            NaN   
8         NaN  2012-06-22        NaN            NaN   
9         NaN  2017-02-01        NaN            NaN   

                                  S_PUESTO D_FEC_INI_PSTO  N_TPO_PSTO  \
0        EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
1                     ASESOR DE NEGOCIOS 2     2018-04-01           9   
2                     ASESOR DE NEGOCIOS 1     2018-10-25           3   
3  ANALISTA SENIOR DE POLITICAS Y PROCESOS     2018-08-29           5   
4                     ASESOR DE NEGOCIOS 3     2014-08-15          53   
5          SUPERVISOR DE BANCA DE SERVICIO     2016-11-01          26   
6                     ASESOR DE NEGOCIOS 1     2018-12-15           1   
7                    EJECUTIVO DE NEGOCIOS     2017-10-21          15   
8             JEFE DE BANCA DE SERVICIO II     2015-04-01          45   
9                     ASESOR DE NEGOCIOS 2     2018-10-25           3   

                         S_DIVISION S_RED_STAFF  
0  DIVISION DE SOPORTE CENTRALIZADO         NaN  
1                          NEGOCIOS         NaN  
2                          NEGOCIOS         NaN  
3                           RIESGOS         NaN  
4                          NEGOCIOS         NaN  
5  DIVISION DE SOPORTE CENTRALIZADO         NaN  
6                          NEGOCIOS         NaN  
7                          NEGOCIOS         NaN  
8  DIVISION DE SOPORTE CENTRALIZADO         NaN  
9                          NEGOCIOS         NaN

In [ ]:
# Filtramos a solo aquellos registros de las divisiones con personal que trabajan con datos.
DIVISIONES = list(puesto_datos['S_DIVISION'].unique())
base1_filtrado = base1_filtrado[(base1_filtrado['S_DIVISION'].isin(DIVISIONES))]
base1_filtrado.head(10)

S_PERIODO  S_MAT S_GENERO D_FEC_NACI  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
0      201901  16537        F 1980-09-27      38     GENERACION X         NaN   
3      201901  11793        M 1983-06-18      35      MILLENNIALS         NaN   
5      201901  12191        M 1977-01-18      41     GENERACION X         NaN   
8      201901  15242        M 1986-10-24      32      MILLENNIALS         NaN   
12     201901  70692        F 1988-07-06      30      MILLENNIALS         NaN   
16     201901  19033        F 1992-05-19      26      MILLENNIALS         NaN   
20     201901  22228        M 1997-01-12      21     GENERACION Z         NaN   
23     201901  63584        F 1984-06-10      34      MILLENNIALS         NaN   
25     201901  64199        M 1984-04-22      34      MILLENNIALS         NaN   
28     201901  60808        M 1983-07-24      35      MILLENNIALS         NaN   

   S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
0          NaN  2013-02-01        NaN            NaN   
3          NaN  2009-05-06        NaN            NaN   
5          NaN  2010-02-01        NaN            NaN   
8          NaN  2012-06-22        NaN            NaN   
12         NaN  2014-10-03        NaN            NaN   
16         NaN  2014-03-12        NaN            NaN   
20         NaN  2018-08-03        NaN            NaN   
23         NaN  2013-09-02        NaN            NaN   
25         NaN  2014-06-02        NaN            NaN   
28         NaN  2007-11-07        NaN            NaN   

                                   S_PUESTO D_FEC_INI_PSTO  N_TPO_PSTO  \
0         EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
3   ANALISTA SENIOR DE POLITICAS Y PROCESOS     2018-08-29           5   
5           SUPERVISOR DE BANCA DE SERVICIO     2016-11-01          26   
8              JEFE DE BANCA DE SERVICIO II     2015-04-01          45   
12    REPRESENTANTE DE BANCA DE SERVICIO II     2016-11-01          26   
16        EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
20    PRACTICANTE DE CONTABILIDAD IMPUESTOS     2018-08-03           5   
23         EJECUTIVO DE BANCA DE SERVICIO I     2018-12-01           1   
25             JEFE DE BANCA DE SERVICIO II     2017-02-15          23   
28              JEFE DE BANCA DE SERVICIO I     2018-05-10           8   

                          S_DIVISION S_RED_STAFF  
0   DIVISION DE SOPORTE CENTRALIZADO         NaN  
3                            RIESGOS         NaN  
5   DIVISION DE SOPORTE CENTRALIZADO         NaN  
8   DIVISION DE SOPORTE CENTRALIZADO         NaN  
12  DIVISION DE SOPORTE CENTRALIZADO         NaN  
16  DIVISION DE SOPORTE CENTRALIZADO         NaN  
20                          FINANZAS         NaN  
23  DIVISION DE SOPORTE CENTRALIZADO         NaN  
25  DIVISION DE SOPORTE CENTRALIZADO         NaN  
28  DIVISION DE SOPORTE CENTRALIZADO         NaN

In [ ]:
#Eliminamos la columna de fecha de nacimiento, ya que ya contamos con la edad.
base1_filtrado = base1_filtrado.drop(columns=['D_FEC_NACI'])
base1_filtrado.head(10)

S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
0      201901  16537        F      38     GENERACION X         NaN   
3      201901  11793        M      35      MILLENNIALS         NaN   
5      201901  12191        M      41     GENERACION X         NaN   
8      201901  15242        M      32      MILLENNIALS         NaN   
12     201901  70692        F      30      MILLENNIALS         NaN   
16     201901  19033        F      26      MILLENNIALS         NaN   
20     201901  22228        M      21     GENERACION Z         NaN   
23     201901  63584        F      34      MILLENNIALS         NaN   
25     201901  64199        M      34      MILLENNIALS         NaN   
28     201901  60808        M      35      MILLENNIALS         NaN   

   S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
0          NaN  2013-02-01        NaN            NaN   
3          NaN  2009-05-06        NaN            NaN   
5          NaN  2010-02-01        NaN            NaN   
8          NaN  2012-06-22        NaN            NaN   
12         NaN  2014-10-03        NaN            NaN   
16         NaN  2014-03-12        NaN            NaN   
20         NaN  2018-08-03        NaN            NaN   
23         NaN  2013-09-02        NaN            NaN   
25         NaN  2014-06-02        NaN            NaN   
28         NaN  2007-11-07        NaN            NaN   

                                   S_PUESTO D_FEC_INI_PSTO  N_TPO_PSTO  \
0         EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
3   ANALISTA SENIOR DE POLITICAS Y PROCESOS     2018-08-29           5   
5           SUPERVISOR DE BANCA DE SERVICIO     2016-11-01          26   
8              JEFE DE BANCA DE SERVICIO II     2015-04-01          45   
12    REPRESENTANTE DE BANCA DE SERVICIO II     2016-11-01          26   
16        EJECUTIVO DE BANCA DE SERVICIO II     2016-11-01          26   
20    PRACTICANTE DE CONTABILIDAD IMPUESTOS     2018-08-03           5   
23         EJECUTIVO DE BANCA DE SERVICIO I     2018-12-01           1   
25             JEFE DE BANCA DE SERVICIO II     2017-02-15          23   
28              JEFE DE BANCA DE SERVICIO I     2018-05-10           8   

                          S_DIVISION S_RED_STAFF  
0   DIVISION DE SOPORTE CENTRALIZADO         NaN  
3                            RIESGOS         NaN  
5   DIVISION DE SOPORTE CENTRALIZADO         NaN  
8   DIVISION DE SOPORTE CENTRALIZADO         NaN  
12  DIVISION DE SOPORTE CENTRALIZADO         NaN  
16  DIVISION DE SOPORTE CENTRALIZADO         NaN  
20                          FINANZAS         NaN  
23  DIVISION DE SOPORTE CENTRALIZADO         NaN  
25  DIVISION DE SOPORTE CENTRALIZADO         NaN  
28  DIVISION DE SOPORTE CENTRALIZADO         NaN

In [ ]:
#Revisamos los registros que contienen el valor "STAFF" en la columna de "S_RED_STAFF".
#Llenamos los puestos con dicho valor nulo
puestos_staff = base1_filtrado['S_PUESTO'][base1_filtrado['S_RED_STAFF']=='STAFF'].dropna().unique()
base1_filtrado.loc[
    (base1_filtrado['S_RED_STAFF'].isna()) & (base1_filtrado['S_PUESTO'].isin(puestos_staff)),
    'S_RED_STAFF'
] = 'STAFF'
print(f"Cantidad total de registros después de los cambios: {len(base1_filtrado)}")

Cantidad total de registros después de los cambios: 193151


In [ ]:
total_registros_staff = len(base1_filtrado[base1_filtrado['S_RED_STAFF'] == 'STAFF'])
print(f"Cantidad total de registros con la categoría STAFF después de los cambios: {total_registros_staff}")

Cantidad total de registros con la categoría STAFF después de los cambios: 38499


In [ ]:
# Contar la cantidad de valores nulos en la columna 'S_RED_STAFF'
nulos_s_red_staff = base1_filtrado['S_RED_STAFF'].isna().sum()

# Contar la cantidad de valores no nulos en la columna 'S_RED_STAFF'
no_nulos_s_red_staff = base1_filtrado['S_RED_STAFF'].notna().sum()

# Ver qué otras categorías están presentes en la columna 'S_RED_STAFF'
categorias_presentes = base1_filtrado['S_RED_STAFF'].dropna().unique()

# Imprimir los resultados
print(f"Cantidad de valores nulos en la columna 'S_RED_STAFF': {nulos_s_red_staff}")
print(f"Cantidad de valores no nulos en la columna 'S_RED_STAFF': {no_nulos_s_red_staff}")
print(f"Otras categorías presentes en la columna 'S_RED_STAFF': {categorias_presentes}")

Cantidad de valores nulos en la columna 'S_RED_STAFF': 154652
Cantidad de valores no nulos en la columna 'S_RED_STAFF': 38499
Otras categorías presentes en la columna 'S_RED_STAFF': ['STAFF']


In [ ]:
#Comenzamos a quedarnos con los puestos de tecnología
PUESTOS_TEC = []
for puesto_tec in base1_filtrado['S_PUESTO'].unique():
  if 'DAT' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'TECN' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'SEGURIDAD' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'SISTEMAS' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif ' TI' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'ANALISTAS' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'ANALISIS' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'INFORMACION' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'SCRUM' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)
  elif 'AGILE' in puesto_tec:
    PUESTOS_TEC.append(puesto_tec)

In [ ]:
print(len(PUESTOS_TEC))
print(len(base1_filtrado['S_PUESTO'].unique()))

249
1117


In [ ]:
PUESTOS_TEC

['ESPECIALISTA DE SOLUCIONES DE NEGOCIO TI',
 'GERENTE DE DESARROLLO DE SISTEMAS',
 'ANALISTA SENIOR DE SEGURIDAD FISICA',
 'AUDITOR DE TI',
 'PRACTICANTE DE AUDITORIA TI',
 'PRACTICANTE DE TECNOLOGIA E INFORMACION DE RIESGOS',
 'AUDITOR SR DE TI',
 'OPERADOR DE SISTEMAS',
 'ANALISTA SENIOR DE INFORMACION TI',
 'GERENTE DE SEGURIDAD FISICA',
 'ANALISTA SENIOR DE GESTION DE INFORMACION Y ANALISIS DE NEGOCIOS',
 'ANALISTA DE SEGURIDAD FISICA',
 'ARQUITECTO ESPECIALISTA TI',
 'SUPERVISOR DE BASE DE DATOS',
 'SUPERVISOR DE SEGURIDAD FISICA',
 'ANALISTA DE SEGURIDAD DE LA INFORMACION',
 'GERENTE DE PRODUCCION DE SISTEMAS',
 'SUPERVISOR DE TECNOLOGIA E INFORMACION DE RIESGOS',
 'PRACTICANTE DE PRODUCCION DE SISTEMAS',
 'ANALISTA SENIOR DE SEGURIDAD DE LA INFORMACION',
 'SUBGERENTE DE SEGURIDAD DE LA INFORMACION',
 'ANALISTA SENIOR DE GESTION DE INFORMACION DE COBRANZA',
 'SUPERVISOR DE ARQUITECTURA TI',
 'GERENTE DE AREA DE SISTEMAS',
 'ANALISTA SENIOR DE SERVICIOS TI',
 'GERENTE DE SERVICIO

### **ANÁLISIS**
Como podemos observar, hay puestos que no son de tecnología, especialmente los que tienen que ver con seguridad física, así que procedemos a omitirlos.

In [ ]:
fisica = []
for puesto in PUESTOS_TEC:
  if 'FISICA' in puesto: fisica.append(puesto)
PUESTOS_TEC = [puesto for puesto in PUESTOS_TEC if puesto not in fisica]
print(len(PUESTOS_TEC)) #PASAMOS DE 248 A 236 PUESTOS

236


In [ ]:
len(base1_filtrado['S_MAT'].unique()) #BASE CON SOLO LAS DIVISIONES CON PUESTOS TECNOLÓGICOS
#NUMERO DE EMPLEADOS

7264

In [ ]:
len(base1_filtrado['S_PUESTO'].unique())

1117

In [ ]:
#CARGAMOS LOS PUESTOS QUE PERTENECEN A LA RED Y ESTÁN EN NUESTRA BASE DE DATOS

puestos_totales = [puesto for puesto in base1_filtrado['S_PUESTO'].unique()]

filtrados_red = base1[base1['S_RED_STAFF'].str.contains('RED', case=False, na=False)]

puestos_en_lista = filtrados_red[filtrados_red['S_PUESTO'].isin(puestos_totales)]

listado_puestos_en_lista = puestos_en_lista['S_PUESTO'].unique()

print("Puestos que están en la categoría 'RED' y en la lista de divisiones:")
for puesto in listado_puestos_en_lista:
    print(puesto)

Puestos que están en la categoría 'RED' y en la lista de divisiones:
ASESOR DE NEGOCIOS 1
EJECUTIVO DE NEGOCIOS
JEFE DE NEGOCIOS
REPRESENTANTE DE BANCA DE SERVICIO II
EJECUTIVO DE RECUPERACIONES I
JEFE DE BANCA DE SERVICIO II
ASESOR DE NEGOCIOS SENIOR
EJECUTIVO DE BANCA DE SERVICIO I
EJECUTIVO DE RECUPERACIONES III
MONITOR DE PRODUCTOS INCLUSIVOS
GERENTE DE AGENCIA
EJECUTIVO DE BANCA DE SERVICIO II
GERENTE REGIONAL
SUBGERENTE DE NEGOCIOS
SUPERVISOR REGIONAL DE RECUPERACIONES
JEFE DE BANCA DE SERVICIO I
REPRESENTANTE DE BANCA DE SERVICIO I - TUTOR
EJECUTIVO DE RECUPERACIONES II
AUXILIAR DE BANCA DE SERVICIO
SUPERVISOR DE BANCA DE SERVICIO
SUPERVISOR REGIONAL DE SOPORTE RED DE AGENCIAS
ASESOR DE INCLUSION FINANCIERA 3
SUPERVISOR DE CUENTAS PASIVAS
EJECUTIVO DE VENTAS DE SERVICIOS
REPRESENTANTE DE BANCA DE SERVICIO TC
SUPERVISOR DE PRODUCTOS INCLUSIVOS
REPRESENTANTE DE BANCA DE SERVICIO PT
EJECUTIVO COMERCIAL
CONSULTOR DE BANCA PRIVADA
ASISTENTE DE NEGOCIOS
ASESOR DE INCLUSION FINANCIERA 

In [ ]:
#Eliminamos puestos que estén en la RED
base1_filtrado = base1_filtrado[~base1_filtrado['S_PUESTO'].isin(listado_puestos_en_lista)]
base1_filtrado.head(10)

S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
3       201901  11793        M      35      MILLENNIALS         NaN   
20      201901  22228        M      21     GENERACION Z         NaN   
40      201901  81856        M      23      MILLENNIALS         NaN   
41      201901  80234        F      25      MILLENNIALS         NaN   
51      201901  78253        M      24      MILLENNIALS         NaN   
54      201901  78548        M      25      MILLENNIALS         NaN   
72      201901  81818        M      33      MILLENNIALS         NaN   
78      201901  82196        M      23      MILLENNIALS         NaN   
91      201901  61290        F      48     GENERACION X         NaN   
101     201901  72762        F      29      MILLENNIALS         NaN   

    S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
3           NaN  2009-05-06        NaN            NaN   
20          NaN  2018-08-03        NaN            NaN   
40          NaN  2018-06-15        NaN            NaN   
41          NaN  2017-12-01        NaN            NaN   
51          NaN  2017-02-03        NaN            NaN   
54          NaN  2017-03-10        NaN            NaN   
72          NaN  2018-06-08        NaN            NaN   
78          NaN  2018-08-10        NaN            NaN   
91          NaN  2009-12-15        NaN            NaN   
101         NaN  2015-06-10        NaN            NaN   

                                            S_PUESTO D_FEC_INI_PSTO  \
3            ANALISTA SENIOR DE POLITICAS Y PROCESOS     2018-08-29   
20             PRACTICANTE DE CONTABILIDAD IMPUESTOS     2018-08-03   
40                          ASESOR TELEFONICO HUB PT     2018-06-15   
41                   ANALISTA DE SERVICIOS CENTRALES     2017-12-01   
51                 OPERADOR JR DE SOPORTE A USUARIOS     2017-02-03   
54   ANALISTA SENIOR DE RIESGO DE MERCADO Y LIQUIDEZ     2018-06-11   
72                                   COORDINADOR HUB     2018-06-08   
78                  ASISTENTE DE SERVICIOS CENTRALES     2018-08-10   
91              GERENTE DE AREA DE RIESGO DE CREDITO     2009-12-15   
101                 ANALISTA CALL CENTER MORA TARDIA     2018-08-10   

     N_TPO_PSTO                        S_DIVISION S_RED_STAFF  
3             5                           RIESGOS       STAFF  
20            5                          FINANZAS       STAFF  
40            7  DIVISION DE SOPORTE CENTRALIZADO         NaN  
41           13  DIVISION DE SOPORTE CENTRALIZADO         NaN  
51           23  DIVISION DE SOPORTE CENTRALIZADO         NaN  
54            7                           RIESGOS       STAFF  
72            7  DIVISION DE SOPORTE CENTRALIZADO         NaN  
78            5  DIVISION DE SOPORTE CENTRALIZADO         NaN  
91          109                           RIESGOS       STAFF  
101           5                           RIESGOS       STAFF

In [ ]:
len(base1_filtrado['S_PUESTO'])

55027

In [ ]:
#Eliminamos la base 1 para no ocupar tanto espacio.
del(base1)

In [ ]:
#Hacemos un conteo de cuantos puestos tuvo cada empleado
print(base1_filtrado.groupby('S_MAT')['S_PUESTO'].nunique())

S_MAT
10011    1
10014    2
10047    1
10179    1
10187    2
        ..
97144    1
97394    1
97395    1
97396    1
97398    1
Name: S_PUESTO, Length: 2249, dtype: int64


In [ ]:
#Guardamos dicho conteo en una lista
conteo_puestos = list(base1_filtrado.groupby('S_MAT')['S_PUESTO'].nunique())
conteo_puestos[0:10]

[1, 2, 1, 1, 2, 1, 3, 3, 2, 1]

In [ ]:
#Obtenemos el primer rango salarial obtenido por cada empleado al ingresar a la empresa,
#borrando duplicados de matriculas y ordenándolos por periodo
primer_sueldo = base1_filtrado.sort_values('S_PERIODO').drop_duplicates('S_MAT',keep='first').reset_index()
primer_sueldo = primer_sueldo.sort_values('S_MAT')
primer_sueldo = primer_sueldo['S_GRD_SALARIAL']
primer_sueldo = primer_sueldo.fillna('A1-STAFF')
print(primer_sueldo)

489             A1-STAFF
394             A1-STAFF
943             A1-STAFF
275             A1-STAFF
677             A1-STAFF
              ...       
2244            A4-STAFF
2245    A4-ESPECIALIZADO
2248    A4-ESPECIALIZADO
2236            A4-STAFF
2240            B1-STAFF
Name: S_GRD_SALARIAL, Length: 2249, dtype: object


In [ ]:
#Filtramos la base de datos al último registro de cada empleado según el último periodo registrado.
base_final = base1_filtrado.sort_values('S_PERIODO').drop_duplicates('S_MAT',keep='last')
base_final = base_final.reset_index()
base_final.head(10)

index  S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION S_EST_CIVIL  \
0   6897     201901  82034        M      26      MILLENNIALS         NaN   
1   7809     201901  80481        M      35      MILLENNIALS         NaN   
2   5543     201901  22284        F      22      MILLENNIALS         NaN   
3   6374     201901  80947        F      20     GENERACION Z         NaN   
4   9825     201901  61311        M      39     GENERACION X         NaN   
5   9901     201901  82789        F      36      MILLENNIALS         NaN   
6   8840     201901  77529        M      38     GENERACION X         NaN   
7   1239     201901  22227        M      25      MILLENNIALS         NaN   
8    528     201901  75701        M      35      MILLENNIALS         NaN   
9     51     201901  78253        M      24      MILLENNIALS         NaN   

  S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
0         NaN  2018-07-11        NaN            NaN   
1         NaN  2018-01-29        NaN            NaN   
2         NaN  2019-01-14        NaN            NaN   
3         NaN  2018-03-02        NaN            NaN   
4         NaN  2010-02-10        NaN            NaN   
5         NaN  2018-10-17        NaN            NaN   
6         NaN  2016-11-14        NaN            NaN   
7         NaN  2018-08-01        NaN            NaN   
8         NaN  2016-04-08        NaN            NaN   
9         NaN  2017-02-03        NaN            NaN   

                                        S_PUESTO D_FEC_INI_PSTO  N_TPO_PSTO  \
0              SUPERVISOR DE SERVICIOS CENTRALES     2018-07-11           6   
1                               AUDITOR SR DE TI     2018-01-29          12   
2                  PRACTICANTE DE TRANSFORMACION     2019-01-14           0   
3           OPERADOR JR DE SOPORTE A USUARIOS PT     2018-03-02          10   
4             ESPECIALISTA DE MODELOS DE RIESGOS     2017-11-01          14   
5                       ASESOR TELEFONICO HUB PT     2018-10-17           3   
6            GERENTE DE SOLUCIONES DE NEGOCIO TI     2016-11-14          26   
7  PRACTICANTE DE AUDITORIA DE RIESGO DE CREDITO     2018-08-01           5   
8                ANALISTA DE SERVICIOS CENTRALES     2016-04-08          33   
9              OPERADOR JR DE SOPORTE A USUARIOS     2017-02-03          23   

                         S_DIVISION S_RED_STAFF  
0  DIVISION DE SOPORTE CENTRALIZADO         NaN  
1                         AUDITORIA         NaN  
2                    TRANSFORMACION       STAFF  
3  DIVISION DE SOPORTE CENTRALIZADO         NaN  
4                           RIESGOS       STAFF  
5  DIVISION DE SOPORTE CENTRALIZADO         NaN  
6  DIVISION DE SOPORTE CENTRALIZADO         NaN  
7                         AUDITORIA       STAFF  
8  DIVISION DE SOPORTE CENTRALIZADO         NaN  
9  DIVISION DE SOPORTE CENTRALIZADO         NaN

In [ ]:
print(len(base_final['S_MAT']))
print(max(base_final['S_PERIODO']))

2249
202406


In [ ]:
#Borramos la columna de STAFF, una vez hechos los filtros necesarios
base_final = base_final.drop(columns=['S_RED_STAFF'])

#Agregamos la columna con el total de conteos que tuvo cada empleado. Pero primero hay que ordenar la base
#Con respecto a la matricula del empleado
base_final = base_final.sort_values('S_MAT')
base_final['TOTAL_PUESTOS'] = conteo_puestos
base_final.head(10)

index  S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION  S_EST_CIVIL  \
2134   52241     202406  10011        F      52     GENERACION X  SOLTERO (A)   
2159   51560     202406  10014        F      58     GENERACION X   CASADO (A)   
1413   40183     202405  10047        M      55     GENERACION X   CASADO (A)   
425    17216     202102  10179        F      44     GENERACION X          NaN   
370   112229     202010  10187        F      42     GENERACION X          NaN   
1584   53896     202406  10280        F      50     GENERACION X  SOLTERO (A)   
1282   36703     202404  10378        F      42      MILLENNIALS  SOLTERO (A)   
2221   51964     202406  10431        M      51     GENERACION X   CASADO (A)   
1869   50197     202406  10454        M      49     GENERACION X  CONVIVIENTE   
1088  112602     202312  10475        F      45     GENERACION X   CASADO (A)   

     S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB S_GRD_SALARIAL  \
2134         NaN  2012-03-01  FULL_TIME       A2-STAFF   
2159         NaN  1999-10-01  FULL_TIME       S2-STAFF   
1413         NaN  1999-03-01        NaN            NaN   
425          NaN  2003-07-07        NaN            NaN   
370          NaN  2001-11-01        NaN            NaN   
1584         NaN  2003-01-01  FULL_TIME       A3-STAFF   
1282         NaN  2003-07-01        NaN            NaN   
2221         NaN  2003-10-01  FULL_TIME       A1-STAFF   
1869         NaN  2003-11-01  FULL_TIME       A1-STAFF   
1088         NaN  2004-01-02        NaN            NaN   

                                               S_PUESTO D_FEC_INI_PSTO  \
2134  SUBGERENTE DE CONTABILIDAD NEGOCIOS Y OPERACIONES     2015-03-01   
2159           GERENTE DE DIVISION DE AUDITORIA INTERNA     1999-10-01   
1413              GERENTE TERRITORIAL DE RECUPERACIONES     2020-02-13   
425            LIDER DE PROYECTOS DE ALIANZAS DIGITALES     2018-12-01   
370             GERENTE DE ESTRATEGIA DE TRANSFORMACION     2019-07-01   
1584                  ESPECIALISTA DE SOPORTE OPERATIVO     2016-02-01   
1282                      ANALISTA SENIOR DE ACTIVOS TI     2023-09-19   
2221                   GERENTE DE AUDITORIA DE AGENCIAS     2022-08-09   
1869    GERENTE DE SUPERVISION Y CONTRALORIA DE RIESGOS     2003-11-01   
1088        GERENTE TERRITORIAL SOPORTE RED DE AGENCIAS     2015-03-01   

      N_TPO_PSTO                   S_DIVISION  TOTAL_PUESTOS  
2134         111         DIVISION DE FINANZAS              1  
2159         296                    AUDITORIA              2  
1413          51                      RIESGOS              1  
425           26               TRANSFORMACION              1  
370           15               TRANSFORMACION              2  
1584         100  TRIBU SEGMENTOS Y PRODUCTOS              1  
1282           7     TECNOLOGIA Y OPERACIONES              3  
2221          22                    AUDITORIA              3  
1869         247                      RIESGOS              2  
1088         105     TECNOLOGIA Y OPERACIONES              1

In [ ]:
# Obtener pares únicos de tipos de puestos y grados salariales
puestos_grados = base_final[['S_PUESTO', 'S_GRD_SALARIAL']].drop_duplicates()

# Imprimir los resultados
print("Relación entre tipos de puestos y grados salariales en la base de datos filtrada:")
for index, row in puestos_grados.iterrows():
    print(f"Puesto: {row['S_PUESTO']}, Grado Salarial: {row['S_GRD_SALARIAL']}")

Relación entre tipos de puestos y grados salariales en la base de datos filtrada:
Puesto: SUBGERENTE DE CONTABILIDAD NEGOCIOS Y OPERACIONES, Grado Salarial: A2-STAFF
Puesto: GERENTE DE DIVISION DE AUDITORIA INTERNA, Grado Salarial: S2-STAFF
Puesto: GERENTE TERRITORIAL DE RECUPERACIONES, Grado Salarial: nan
Puesto: LIDER DE PROYECTOS DE ALIANZAS DIGITALES, Grado Salarial: nan
Puesto: GERENTE DE ESTRATEGIA DE TRANSFORMACION, Grado Salarial: nan
Puesto: ESPECIALISTA DE SOPORTE OPERATIVO, Grado Salarial: A3-STAFF
Puesto: ANALISTA SENIOR DE ACTIVOS TI, Grado Salarial: nan
Puesto: GERENTE DE AUDITORIA DE AGENCIAS, Grado Salarial: A1-STAFF
Puesto: GERENTE DE SUPERVISION Y CONTRALORIA DE RIESGOS, Grado Salarial: A1-STAFF
Puesto: GERENTE TERRITORIAL SOPORTE RED DE AGENCIAS, Grado Salarial: nan
Puesto: GERENTE DE AREA DE SISTEMAS, Grado Salarial: nan
Puesto: SUBGERENTE DE POLITICAS Y PROCESOS, Grado Salarial: nan
Puesto: SUBGERENTE DE RIESGO OPERATIVO, Grado Salarial: A2-STAFF
Puesto: ESPECIALIS

In [ ]:
# Obtener los grados salariales únicos
grados_salariales_unicos = base_final['S_GRD_SALARIAL'].dropna().unique()

# Imprimir los grados salariales únicos
print("Grados salariales únicos en la base de datos:")
for grado in grados_salariales_unicos:
    print(grado)

Grados salariales únicos en la base de datos:
A2-STAFF
S2-STAFF
A3-STAFF
A1-STAFF
S3-STAFF
A4-STAFF
A2-ESPECIALIZADO
S2-ESPECIALISTA
S3-ESPECIALIZADO
A5-STAFF
A1-ESPECIALIZADO
A3-ESPECIALIZADO
A2-ESPECIALISTA
A1-ESPECIALISTA
S3-ESPECIALISTA
S2-XPERT
S2-CORE TEAM
A4-ESPECIALIZADO
S3-CORE TEAM
A3-CONSULTORIA
A4-CONSULTORIA
S3-CONSULTORIA
A2-CONSULTORIA
A5-CONSULTORIA
S3-HIPO
A5-ESPECIALIZADO
A1-CONSULTORIA
B1-STAFF


In [ ]:
# Crear un diccionario de rangos salariales según el grado salarial

rango_salarial = {
    'A1-STAFF': (2000, 3000),
    'A2-STAFF': (3000, 4000),
    'A3-STAFF': (4000, 5000),
    'A4-STAFF': (5000, 6000),
    'A5-STAFF': (6000, 7000),
    'A1-ESPECIALIZADO': (2500, 3500),
    'A2-ESPECIALIZADO': (3500, 4500),
    'A3-ESPECIALIZADO': (4500, 5500),
    'A4-ESPECIALIZADO': (5500, 6500),
    'A5-ESPECIALIZADO': (6500, 7500),
    'S2-STAFF': (7500, 8500),
    'S3-STAFF': (8500, 9500),
    'S2-ESPECIALISTA': (9000, 10000),
    'S3-ESPECIALISTA': (10000, 11000),
    'A1-CONSULTORIA': (3000, 4000),
    'A2-CONSULTORIA': (4000, 5000),
    'A3-CONSULTORIA': (5000, 6000),
    'A4-CONSULTORIA': (6000, 7000),
    'A5-CONSULTORIA': (7000, 8000),
    'S3-CORE TEAM': (12000, 13000),
    'S2-CORE TEAM': (11000, 12000),
    'B1-STAFF': (5000, 6000),
    'S3-HIPO': (14000, 15000),
    'S2-XPERT': (13000, 14000),
    # Añadir grados salariales adicionales según sea necesario
    'S3-ESPECIALIZADO': (10000, 12000),
    'S3-CONSULTORIA': (11000, 13000),
    'A1-ESPECIALISTA': (3000, 4000),
    'A2-ESPECIALISTA': (3500, 4500),
    'A3-ESPECIALISTA': (4000, 5000),
    'A4-ESPECIALISTA': (4500, 5500),
    'A5-ESPECIALISTA': (5000, 6000),
    # Añadir más grados salariales aquí si es necesario
}

# Crear un DataFrame con rangos mínimos y máximos para cada grado salarial
rango_minimo = {key: val[0] for key, val in rango_salarial.items()}
rango_maximo = {key: val[1] for key, val in rango_salarial.items()}

# Rellenar valores nulos en 'S_GRD_SALARIAL' con un valor predeterminado si es necesario
base_final['S_GRD_SALARIAL'].fillna('A1-STAFF', inplace=True)

# Verificar si todos los grados salariales están en el diccionario
def verificar_grado_salarial(grado_salarial):
    if grado_salarial not in rango_minimo:
        print(f"Grado salarial '{grado_salarial}' no está en el diccionario de rangos.")
        return 'A1-STAFF'
    return grado_salarial

base_final['S_GRD_SALARIAL'] = base_final['S_GRD_SALARIAL'].apply(verificar_grado_salarial)

# Asignar un rango salarial aleatorio basado en el grado salarial
def asignar_rango_aleatorio(grado_salarial):
    if grado_salarial in rango_minimo:
        min_salario = rango_minimo[grado_salarial]
        max_salario = rango_maximo[grado_salarial]
        return np.random.randint(min_salario, max_salario + 1)
    return np.nan

base_final['RANG_SALARIAL'] = base_final['S_GRD_SALARIAL'].apply(asignar_rango_aleatorio)
base_final['RANG_SALARIAL_INICIAL'] = primer_sueldo.apply(asignar_rango_aleatorio)


# Ordenar el DataFrame por 'S_GRD_SALARIAL' (S primero, A después)
base_final = base_final.sort_values(by='S_GRD_SALARIAL', ascending=False)

In [ ]:
# Verificar la cantidad de valores nulos en la columna 'S_GRD_SALARIAL'
num_nulos = base_final['RANG_SALARIAL'].isna().sum()

print(f"Número de valores nulos en la columna 'S_GRD_SALARIAL': {num_nulos}")

Número de valores nulos en la columna 'S_GRD_SALARIAL': 0


In [ ]:
# Verificar la cantidad de valores nulos en la columna 'RANG_SALARIAL_INICIAL'
num_nulos = base_final['RANG_SALARIAL_INICIAL'].isna().sum()

print(f"Número de valores nulos en la columna 'RANG_SALARIAL_INICIAL': {num_nulos}")

Número de valores nulos en la columna 'RANG_SALARIAL_INICIAL': 0


In [ ]:
# Contar el número total de registros en el DataFrame
num_registros = len(base_final)

print(f"Número total de registros en el DataFrame: {num_registros}")

Número total de registros en el DataFrame: 2249


In [ ]:
# Listado de tipos de contrato disponibles y sus frecuencias
frecuencias_contrato = {
    'INDETERMINADO': 241,
    'A PLAZO INDET - D.LEG.': 206,
    'PERS. CONVENIO DE PRACTICAS PREPROFESIONALES': 55,
    'PERS. CON CONVENIO DE PRACTICAS PROFESIONALES': 5,
    'DE SUPLENCIA': 3,
    'POR INICIO O INCREM DE ACTIV': 1
}

# Crear una lista de contratos repetidos según sus frecuencias
contratos_list = []
for contrato, freq in frecuencias_contrato.items():
    contratos_list.extend([contrato] * freq)

# Verificar todos los tipos de puesto únicos en el DataFrame
tipos_puesto_unicos = base_final['S_PUESTO'].unique()
print("Tipos de puesto únicos en el DataFrame:", tipos_puesto_unicos)

# Asignar contratos aleatorios para los nulos en 'S_TIP_CTRTO' de forma equilibrada
def asignar_contrato_equilibrado(num_nulos):
    if len(contratos_list) == 0:
        return 'INDETERMINADO'
    return np.random.choice(contratos_list)

# Filtrar los registros con valores nulos en 'S_TIP_CTRTO'
nulos_contrato = base_final[base_final['S_TIP_CTRTO'].isna()]

# Asignar contratos equilibrados a los registros con valores nulos en 'S_TIP_CTRTO'
base_final.loc[base_final['S_TIP_CTRTO'].isna(), 'S_TIP_CTRTO'] = [asignar_contrato_equilibrado(len(nulos_contrato)) for _ in range(len(nulos_contrato))]

# Verificar el resultado
num_nulos_final = base_final['S_TIP_CTRTO'].isna().sum()
print("Número de nulos en 'S_TIP_CTRTO' después de asignar contratos:", num_nulos_final)

# Verificar la distribución final de los contratos
print(base_final['S_TIP_CTRTO'].value_counts())

Tipos de puesto únicos en el DataFrame: ['TRIBE TECHNICAL LEADER' 'GERENTE DE SERVICIO DE CANALES DE COBRANZA'
 'GERENTE DE SERVICIO DE CONTROL FINANCIERO'
 'GERENTE DE SERVICIO DE SOSTENIBILIDAD E INCLUSION FINANCIERA'
 'GERENTE DE SERVICIO DE RIESGOS DE TESORERIA'
 'PRODUCT OWNER DE ESTRATEGIA MORA TARDIA DE RECUPERACIONES'
 'PRODUCT OWNER DE PLANEAMIENTO Y DATA'
 'GERENTE DE SERVICIO DE GESTION DE LA DEMANDA DE TI'
 'PRODUCT OWNER DE CRM' 'GERENTE DE SERVICIO DE GOBIERNO DE DATOS'
 'GERENTE DE SERVICIO DE AUDITORIA DE PROCESOS Y PLANEAMIENTO'
 'GERENTE DE SERVICIO DE PAUTA Y PROCESOS'
 'GERENTE DE SERVICIO DE OPERACIONES DE SEGURIDAD TECNOLOGICA'
 'GERENTE DE SERVICIO DE ESTRATEGIA DE RECUPERACIONES'
 'GERENTE DE SERVICIO DE TRIBU DE DATOS'
 'GERENTE DE SERVICIO DE ANALITICA AVANZADA'
 'GERENTE DE SERVICIO DE TRIBU DE DATOS CLOUD'
 'GERENTE DE SERVICIO DE MODELOS DE RIESGOS'
 'GERENTE DE SERVICIO DE ARQUITECTURA EMPRESARIAL'
 'GERENTE DE SERVICIO DE GESTION DE BALANCE'
 'GERENTE DE 

In [ ]:
base_final['S_TIP_CTRTO'].value_counts()

S_TIP_CTRTO
INDETERMINADO                                    1070
A PLAZO INDET - D.LEG.                            699
PERS. CONVENIO DE PRACTICAS PREPROFESIONALES      235
A PLAZO INDET - D.LEG. 728                        206
PERS. CON CONVENIO DE PRACTICAS PROFESIONALES      18
DE SUPLENCIA                                       17
POR INICIO O INCREM DE ACTIV                        4
Name: count, dtype: int64

In [ ]:
#ELiminamos la variable de grado salarial. Nos quedamos solo con la columna de rango salarial
base_final = base_final.drop(columns=['S_GRD_SALARIAL'])
base_final.head(10)

index  S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION  S_EST_CIVIL  \
1627  54435     202406  11060        F      49     GENERACION X  SOLTERO (A)   
1716  53680     202406  12682        F      41      MILLENNIALS   CASADO (A)   
2166  51612     202406  87728        M      37      MILLENNIALS   CASADO (A)   
2155  51499     202406  62270        F      46     GENERACION X   CASADO (A)   
1923  50509     202406  11423        F      43      MILLENNIALS  SOLTERO (A)   
1735  58377     202406  82105        M      31      MILLENNIALS  SOLTERO (A)   
1993  48968     202406  89420        F      38      MILLENNIALS   CASADO (A)   
1701  58622     202406  97139        M      33      MILLENNIALS   CASADO (A)   
2113  53539     202406  89458        F      39      MILLENNIALS   CASADO (A)   
1748  56492     202406  89975        M      41      MILLENNIALS   CASADO (A)   

                                       S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB  \
1627                                 INDETERMINADO  2007-03-26  FULL_TIME   
1716  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES  2010-09-20  FULL_TIME   
2166                        A PLAZO INDET - D.LEG.  2019-12-01  FULL_TIME   
2155                                 INDETERMINADO  2012-02-15  FULL_TIME   
1923                                 INDETERMINADO  2008-07-21  FULL_TIME   
1735                                  DE SUPLENCIA  2018-07-25  FULL_TIME   
1993                        A PLAZO INDET - D.LEG.  2021-02-01  FULL_TIME   
1701                    A PLAZO INDET - D.LEG. 728  2024-06-01  FULL_TIME   
2113                                 INDETERMINADO  2021-02-15  FULL_TIME   
1748                        A PLAZO INDET - D.LEG.  2021-10-22  FULL_TIME   

                                               S_PUESTO D_FEC_INI_PSTO  \
1627                             TRIBE TECHNICAL LEADER     2023-09-15   
1716         GERENTE DE SERVICIO DE CANALES DE COBRANZA     2024-04-19   
2166          GERENTE DE SERVICIO DE CONTROL FINANCIERO     2019-12-01   
2155  GERENTE DE SERVICIO DE SOSTENIBILIDAD E INCLUS...     2023-07-04   
1923        GERENTE DE SERVICIO DE RIESGOS DE TESORERIA     2016-07-11   
1735  PRODUCT OWNER DE ESTRATEGIA MORA TARDIA DE REC...     2024-03-07   
1993               PRODUCT OWNER DE PLANEAMIENTO Y DATA     2022-04-04   
1701  GERENTE DE SERVICIO DE GESTION DE LA DEMANDA D...     2024-06-01   
2113                               PRODUCT OWNER DE CRM     2021-04-22   
1748                             TRIBE TECHNICAL LEADER     2021-10-22   

      N_TPO_PSTO                                         S_DIVISION  \
1627           9                                         TECNOLOGIA   
1716           2                                            RIESGOS   
2166          54                               DIVISION DE FINANZAS   
2155          11  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
1923          95                                            RIESGOS   
1735           3                                            RIESGOS   
1993          26                        TRIBU SEGMENTOS Y PRODUCTOS   
1701           0                                         TECNOLOGIA   
2113          38  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
1748          32                                         TECNOLOGIA   

      TOTAL_PUESTOS  RANG_SALARIAL  RANG_SALARIAL_INICIAL  
1627              3           8947                   2563  
1716              3           8516                   2801  
2166              1           8701                   2877  
2155              4           9073                   2333  
1923              1           8756                   2978  
1735              3           9318                   2078  
1993              2           8644                   2034  
1701              1           8597                   2337  
2113              1           9069                   2820  
1748              1           9298                   2700

se verifican la cantidad de valores nulos en la columna de estado civil

In [ ]:
tiene_nulos = base_final['S_EST_CIVIL'].isnull().any()
print(f"¿La columna 'S_EST_CIVIL' tiene nulos? {tiene_nulos}")

# Contar la cantidad de valores nulos en la columna
cantidad_nulos = base_final['S_EST_CIVIL'].isnull().sum()
print(f"Cantidad de valores nulos en 'nombre_columna': {cantidad_nulos}")

¿La columna 'S_EST_CIVIL' tiene nulos? True
Cantidad de valores nulos en 'nombre_columna': 1051


Creamos un codigo que define grupos de edad, se analiza la distribucion de las categorias y se crea una funcion para ya poner el estado civil basados en la distribucion y el grupo de edad

In [ ]:
# Definimos los grupos de edad
bins = [18, 30, 40, 50, 100]
labels = ['Joven', 'Adulto', 'Maduro', 'Mayor']
base_final['Grupo_Edad'] = pd.cut(base_final['N_EDAD'], bins=bins, labels=labels, right=False)

# Analizamos la distribución de estado civil en cada grupo de edad
distribucion_est_civil = base_final.groupby('Grupo_Edad')['S_EST_CIVIL'].value_counts(normalize=True).unstack().fillna(0)

# Función para imputar estado civil basado en la distribución y el grupo de edad
def imputar_estado_civil(grupo_edad, distribucion):
    if pd.isna(grupo_edad):
        return np.nan
    return np.random.choice(distribucion.columns, p=distribucion.loc[grupo_edad].values)

# Imputación de valores faltantes en la columna 'S_EST_CIVIL'
base_final['S_EST_CIVIL'] = base_final.apply(
    lambda row: imputar_estado_civil(row['Grupo_Edad'], distribucion_est_civil) if pd.isna(row['S_EST_CIVIL']) else row['S_EST_CIVIL'],
    axis=1
)

# Eliminamos la columna 'Grupo_Edad' ya que no queremos columnas adicionales
base_final.drop(columns=['Grupo_Edad'], inplace=True)

# Verificamos si aún existen nulos
nulos_restantes = base_final['S_EST_CIVIL'].isnull().sum()

<ipython-input-44-6a110728f02c>:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  distribucion_est_civil = base_final.groupby('Grupo_Edad')['S_EST_CIVIL'].value_counts(normalize=True).unstack().fillna(0)


In [ ]:
tiene_nulos = base_final['S_EST_CIVIL'].isnull().any()
print(f"¿La columna 'S_EST_CIVIL' tiene nulos? {tiene_nulos}")

# Contar la cantidad de valores nulos en la columna
cantidad_nulos = base_final['S_EST_CIVIL'].isnull().sum()
print(f"Cantidad de valores nulos en 'nombre_columna': {cantidad_nulos}")

¿La columna 'S_EST_CIVIL' tiene nulos? False
Cantidad de valores nulos en 'nombre_columna': 0


### Como definitiva nos hemos quedado con 2249 empleados activos

Ya que la columna de jornada laboral solo tiene full time determinamos que todas las jornadas quedan como full time si desean asignarlo de manera diferente se aceptan sugerencias

In [ ]:
# Asignar 'FULL_TIME' a todos los valores nulos en 'S_JRND_LAB'
base_final['S_JRND_LAB'].fillna('FULL_TIME', inplace=True)

# Verificar que ya no haya nulos en 'S_JRND_LAB'
nulos_jornada = base_final['S_JRND_LAB'].isnull().sum()

print(f"Número de nulos en 'S_JRND_LAB' después de la asignación: {nulos_jornada}")
print(base_final['S_EST_CIVIL'].value_counts())

Número de nulos en 'S_JRND_LAB' después de la asignación: 0
S_EST_CIVIL
SOLTERO (A)       1623
CASADO (A)         476
CONVIVIENTE        108
DIVORCIADO (A)      41
VIUDO (A)            1
Name: count, dtype: int64


In [ ]:
base_final.head(10)

index  S_PERIODO  S_MAT S_GENERO  N_EDAD S_GRP_GENERACION  S_EST_CIVIL  \
1627  54435     202406  11060        F      49     GENERACION X  SOLTERO (A)   
1716  53680     202406  12682        F      41      MILLENNIALS   CASADO (A)   
2166  51612     202406  87728        M      37      MILLENNIALS   CASADO (A)   
2155  51499     202406  62270        F      46     GENERACION X   CASADO (A)   
1923  50509     202406  11423        F      43      MILLENNIALS  SOLTERO (A)   
1735  58377     202406  82105        M      31      MILLENNIALS  SOLTERO (A)   
1993  48968     202406  89420        F      38      MILLENNIALS   CASADO (A)   
1701  58622     202406  97139        M      33      MILLENNIALS   CASADO (A)   
2113  53539     202406  89458        F      39      MILLENNIALS   CASADO (A)   
1748  56492     202406  89975        M      41      MILLENNIALS   CASADO (A)   

                                       S_TIP_CTRTO D_FEC_CTRTO S_JRND_LAB  \
1627                                 INDETERMINADO  2007-03-26  FULL_TIME   
1716  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES  2010-09-20  FULL_TIME   
2166                        A PLAZO INDET - D.LEG.  2019-12-01  FULL_TIME   
2155                                 INDETERMINADO  2012-02-15  FULL_TIME   
1923                                 INDETERMINADO  2008-07-21  FULL_TIME   
1735                                  DE SUPLENCIA  2018-07-25  FULL_TIME   
1993                        A PLAZO INDET - D.LEG.  2021-02-01  FULL_TIME   
1701                    A PLAZO INDET - D.LEG. 728  2024-06-01  FULL_TIME   
2113                                 INDETERMINADO  2021-02-15  FULL_TIME   
1748                        A PLAZO INDET - D.LEG.  2021-10-22  FULL_TIME   

                                               S_PUESTO D_FEC_INI_PSTO  \
1627                             TRIBE TECHNICAL LEADER     2023-09-15   
1716         GERENTE DE SERVICIO DE CANALES DE COBRANZA     2024-04-19   
2166          GERENTE DE SERVICIO DE CONTROL FINANCIERO     2019-12-01   
2155  GERENTE DE SERVICIO DE SOSTENIBILIDAD E INCLUS...     2023-07-04   
1923        GERENTE DE SERVICIO DE RIESGOS DE TESORERIA     2016-07-11   
1735  PRODUCT OWNER DE ESTRATEGIA MORA TARDIA DE REC...     2024-03-07   
1993               PRODUCT OWNER DE PLANEAMIENTO Y DATA     2022-04-04   
1701  GERENTE DE SERVICIO DE GESTION DE LA DEMANDA D...     2024-06-01   
2113                               PRODUCT OWNER DE CRM     2021-04-22   
1748                             TRIBE TECHNICAL LEADER     2021-10-22   

      N_TPO_PSTO                                         S_DIVISION  \
1627           9                                         TECNOLOGIA   
1716           2                                            RIESGOS   
2166          54                               DIVISION DE FINANZAS   
2155          11  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
1923          95                                            RIESGOS   
1735           3                                            RIESGOS   
1993          26                        TRIBU SEGMENTOS Y PRODUCTOS   
1701           0                                         TECNOLOGIA   
2113          38  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
1748          32                                         TECNOLOGIA   

      TOTAL_PUESTOS  RANG_SALARIAL  RANG_SALARIAL_INICIAL  
1627              3           8947                   2563  
1716              3           8516                   2801  
2166              1           8701                   2877  
2155              4           9073                   2333  
1923              1           8756                   2978  
1735              3           9318                   2078  
1993              2           8644                   2034  
1701              1           8597                   2337  
2113              1           9069                   2820  
1748              1           9298                   2700

Por ultimo verificamos si existen valores nulos en cualquiera de las columnas de la base final

In [ ]:
# Verificar la cantidad de nulos por columna
nulos_por_columna = base_final.isnull().sum()

# Mostrar las columnas que tienen nulos (si hay)
nulos_presentes = nulos_por_columna[nulos_por_columna > 0]

if nulos_presentes.empty:
    print("No hay valores nulos en ninguna columna.")
else:
    print("Hay valores nulos en las siguientes columnas:")
    print(nulos_presentes)

No hay valores nulos en ninguna columna.


### **RENOMBRADO DE VARIABLES**
A partir de este punto renombraremos las variables de tal forma que puedan ser mayormente reconocidas por el equipo y quien le interese investigar la base.

In [ ]:
#Comenzamos con el renombre de columnas en la base de datos
base_final = base_final.rename(columns={'S_MAT': 'MATRICULA',
                           'S_PUESTO': 'PUESTO',
                           'S_DIVISION': 'DIVISION',
                           'N_EDAD': 'EDAD',
                           'S_GENERO': 'GENERO',
                           'S_PERIODO': 'PERIODO',
                           'S_FECHA': 'FECHA',
                           'S_GRP_GENERACION': 'GRUPO_GENERACION',
                           'S_TIP_CTRTO': 'TIPO_CONTRATO',
                           'S_EST_CIVIL': 'ESTADO_CIVIL',
                           'D_FEC_CTRTO': 'FECHA_CONTRATACION',
                           'D_FEC_INI_PSTO': 'FECHA_INICIO_PUESTO',
                           'N_TPO_PSTO': 'TIEMPO_PUESTO',
                           'S_JRND_LAB': 'JORNADA_LABORAL',
                           'RANG_SALARIAL': 'RANGO_SALARIAL'
})
base_final.head(10)




index  PERIODO  MATRICULA GENERO  EDAD GRUPO_GENERACION ESTADO_CIVIL  \
1627  54435   202406      11060      F    49     GENERACION X  SOLTERO (A)   
1716  53680   202406      12682      F    41      MILLENNIALS   CASADO (A)   
2166  51612   202406      87728      M    37      MILLENNIALS   CASADO (A)   
2155  51499   202406      62270      F    46     GENERACION X   CASADO (A)   
1923  50509   202406      11423      F    43      MILLENNIALS  SOLTERO (A)   
1735  58377   202406      82105      M    31      MILLENNIALS  SOLTERO (A)   
1993  48968   202406      89420      F    38      MILLENNIALS   CASADO (A)   
1701  58622   202406      97139      M    33      MILLENNIALS   CASADO (A)   
2113  53539   202406      89458      F    39      MILLENNIALS   CASADO (A)   
1748  56492   202406      89975      M    41      MILLENNIALS   CASADO (A)   

                                     TIPO_CONTRATO FECHA_CONTRATACION  \
1627                                 INDETERMINADO         2007-03-26   
1716  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES         2010-09-20   
2166                        A PLAZO INDET - D.LEG.         2019-12-01   
2155                                 INDETERMINADO         2012-02-15   
1923                                 INDETERMINADO         2008-07-21   
1735                                  DE SUPLENCIA         2018-07-25   
1993                        A PLAZO INDET - D.LEG.         2021-02-01   
1701                    A PLAZO INDET - D.LEG. 728         2024-06-01   
2113                                 INDETERMINADO         2021-02-15   
1748                        A PLAZO INDET - D.LEG.         2021-10-22   

     JORNADA_LABORAL                                             PUESTO  \
1627       FULL_TIME                             TRIBE TECHNICAL LEADER   
1716       FULL_TIME         GERENTE DE SERVICIO DE CANALES DE COBRANZA   
2166       FULL_TIME          GERENTE DE SERVICIO DE CONTROL FINANCIERO   
2155       FULL_TIME  GERENTE DE SERVICIO DE SOSTENIBILIDAD E INCLUS...   
1923       FULL_TIME        GERENTE DE SERVICIO DE RIESGOS DE TESORERIA   
1735       FULL_TIME  PRODUCT OWNER DE ESTRATEGIA MORA TARDIA DE REC...   
1993       FULL_TIME               PRODUCT OWNER DE PLANEAMIENTO Y DATA   
1701       FULL_TIME  GERENTE DE SERVICIO DE GESTION DE LA DEMANDA D...   
2113       FULL_TIME                               PRODUCT OWNER DE CRM   
1748       FULL_TIME                             TRIBE TECHNICAL LEADER   

     FECHA_INICIO_PUESTO  TIEMPO_PUESTO  \
1627          2023-09-15              9   
1716          2024-04-19              2   
2166          2019-12-01             54   
2155          2023-07-04             11   
1923          2016-07-11             95   
1735          2024-03-07              3   
1993          2022-04-04             26   
1701          2024-06-01              0   
2113          2021-04-22             38   
1748          2021-10-22             32   

                                               DIVISION  TOTAL_PUESTOS  \
1627                                         TECNOLOGIA              3   
1716                                            RIESGOS              3   
2166                               DIVISION DE FINANZAS              1   
2155  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...              4   
1923                                            RIESGOS              1   
1735                                            RIESGOS              3   
1993                        TRIBU SEGMENTOS Y PRODUCTOS              2   
1701                                         TECNOLOGIA              1   
2113  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...              1   
1748                                         TECNOLOGIA              1   

      RANGO_SALARIAL  RANG_SALARIAL_INICIAL  
1627            8947                   2563  
1716            8516                   2801  
2166            8701                   2877  
2155            9073                   2333 

### **Anexamos base de Cesados**

In [ ]:
# URL del archivo CSV en AWS
url = 'https://base-project-g14-iebs.s3.amazonaws.com/CL.xlsx'

# Cargar el CSV en un DataFrame
base_cesados = pd.read_excel(url)

base_cesados.head(10)

S_PERIODO  S_MAT D_FEC_CESE   S_DEC_CESE                S_MTV_CESE  \
0     201901  71800 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA   
1     201901  72077 2019-01-14  COLABORADOR     RENUNCIA - VOLUNTARIA   
2     201901  71551 2019-01-03  COLABORADOR     RENUNCIA - VOLUNTARIA   
3     201901  71221 2019-01-30  COLABORADOR     RENUNCIA - VOLUNTARIA   
4     201901  71532 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA   
5     201901  72898 2019-01-31        BANCO  RENUNCIA - MUTUO DISENSO   
6     201901  73259 2019-01-02        BANCO  RENUNCIA - MUTUO DISENSO   
7     201901  72835 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA   
8     201901  72093 2019-01-07  COLABORADOR     RENUNCIA - VOLUNTARIA   
9     201901  72446 2019-01-04  COLABORADOR     RENUNCIA - VOLUNTARIA   

  S_SUB_MTV_CESE S_MTV_CESE_CAT S_SUB_MTV_CESE_CAT  \
0       RENUNCIA            NaN                NaN   
1       RENUNCIA            NaN                NaN   
2       RENUNCIA            NaN                NaN   
3       RENUNCIA            NaN                NaN   
4       RENUNCIA            NaN                NaN   
5  MUTUO ACUERDO            NaN                NaN   
6  MUTUO ACUERDO            NaN                NaN   
7       RENUNCIA            NaN                NaN   
8       RENUNCIA            NaN                NaN   
9       RENUNCIA            NaN                NaN   

                                          S_OBS_CESE  
0  Asesora tiene un oferta de otra entidad con me...  
1                                 MOTIVOS PERSONALES  
2                                RENUNCIA VOLUNTARIA  
3                      RENUNCIA POR MOTIVOS DE SALUD  
4                                 MOTIVOS PERSONALES  
5                                   Perfil de puesto  
6                                   PERFIL - LÍDERES  
7  OFERTA DE MAYOR SUELDO POR PARTE DE CAJA AREQU...  
8                                 MOTIVOS PERSONALES  
9  Renuncia voluntaria por motivos personales (fa...

In [ ]:
base_cesados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14818 entries, 0 to 14817
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   S_PERIODO           14818 non-null  int64         
 1   S_MAT               14818 non-null  int64         
 2   D_FEC_CESE          14818 non-null  datetime64[ns]
 3   S_DEC_CESE          14817 non-null  object        
 4   S_MTV_CESE          14812 non-null  object        
 5   S_SUB_MTV_CESE      14683 non-null  object        
 6   S_MTV_CESE_CAT      9379 non-null   object        
 7   S_SUB_MTV_CESE_CAT  9539 non-null   object        
 8   S_OBS_CESE          13965 non-null  object        
dtypes: datetime64[ns](1), int64(2), object(6)
memory usage: 1.0+ MB


In [ ]:
#Corroboramos cuantos empleados aparecen en la base de cesados.
base_cesados['S_MAT'][base_cesados['S_MAT'].isin(base_final['MATRICULA'])].nunique()

1076

In [ ]:
len(base_final['MATRICULA'])

2249

In [ ]:
#Eliminamos columnas innecesarias de la base de cesados
base_final_cesados = base_cesados.drop(columns=['S_MTV_CESE_CAT','S_SUB_MTV_CESE_CAT','S_PERIODO'],axis = 1)
base_final_cesados.head(10)

S_MAT D_FEC_CESE   S_DEC_CESE                S_MTV_CESE S_SUB_MTV_CESE  \
0  71800 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
1  72077 2019-01-14  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
2  71551 2019-01-03  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
3  71221 2019-01-30  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
4  71532 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
5  72898 2019-01-31        BANCO  RENUNCIA - MUTUO DISENSO  MUTUO ACUERDO   
6  73259 2019-01-02        BANCO  RENUNCIA - MUTUO DISENSO  MUTUO ACUERDO   
7  72835 2019-01-02  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
8  72093 2019-01-07  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   
9  72446 2019-01-04  COLABORADOR     RENUNCIA - VOLUNTARIA       RENUNCIA   

                                          S_OBS_CESE  
0  Asesora tiene un oferta de otra entidad con me...  
1                                 MOTIVOS PERSONALES  
2                                RENUNCIA VOLUNTARIA  
3                      RENUNCIA POR MOTIVOS DE SALUD  
4                                 MOTIVOS PERSONALES  
5                                   Perfil de puesto  
6                                   PERFIL - LÍDERES  
7  OFERTA DE MAYOR SUELDO POR PARTE DE CAJA AREQU...  
8                                 MOTIVOS PERSONALES  
9  Renuncia voluntaria por motivos personales (fa...

In [ ]:
#HACEMOS MERGE DE LA BASE FINAL CON LA DE CESADOS
base_final_merge = pd.merge(base_final, base_final_cesados, left_on='MATRICULA', right_on = 'S_MAT',how='left')
base_final_merge.head(10)

index  PERIODO  MATRICULA GENERO  EDAD GRUPO_GENERACION ESTADO_CIVIL  \
0  54435   202406      11060      F    49     GENERACION X  SOLTERO (A)   
1  53680   202406      12682      F    41      MILLENNIALS   CASADO (A)   
2  51612   202406      87728      M    37      MILLENNIALS   CASADO (A)   
3  51499   202406      62270      F    46     GENERACION X   CASADO (A)   
4  50509   202406      11423      F    43      MILLENNIALS  SOLTERO (A)   
5  58377   202406      82105      M    31      MILLENNIALS  SOLTERO (A)   
6  48968   202406      89420      F    38      MILLENNIALS   CASADO (A)   
7  58622   202406      97139      M    33      MILLENNIALS   CASADO (A)   
8  53539   202406      89458      F    39      MILLENNIALS   CASADO (A)   
9  56492   202406      89975      M    41      MILLENNIALS   CASADO (A)   

                                  TIPO_CONTRATO FECHA_CONTRATACION  \
0                                 INDETERMINADO         2007-03-26   
1  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES         2010-09-20   
2                        A PLAZO INDET - D.LEG.         2019-12-01   
3                                 INDETERMINADO         2012-02-15   
4                                 INDETERMINADO         2008-07-21   
5                                  DE SUPLENCIA         2018-07-25   
6                        A PLAZO INDET - D.LEG.         2021-02-01   
7                    A PLAZO INDET - D.LEG. 728         2024-06-01   
8                                 INDETERMINADO         2021-02-15   
9                        A PLAZO INDET - D.LEG.         2021-10-22   

  JORNADA_LABORAL  ...                                           DIVISION  \
0       FULL_TIME  ...                                         TECNOLOGIA   
1       FULL_TIME  ...                                            RIESGOS   
2       FULL_TIME  ...                               DIVISION DE FINANZAS   
3       FULL_TIME  ...  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
4       FULL_TIME  ...                                            RIESGOS   
5       FULL_TIME  ...                                            RIESGOS   
6       FULL_TIME  ...                        TRIBU SEGMENTOS Y PRODUCTOS   
7       FULL_TIME  ...                                         TECNOLOGIA   
8       FULL_TIME  ...  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...   
9       FULL_TIME  ...                                         TECNOLOGIA   

  TOTAL_PUESTOS  RANGO_SALARIAL RANG_SALARIAL_INICIAL  S_MAT  D_FEC_CESE  \
0             3            8947                  2563    NaN         NaT   
1             3            8516                  2801    NaN         NaT   
2             1            8701                  2877    NaN         NaT   
3             4            9073                  2333    NaN         NaT   
4             1            8756                  2978    NaN         NaT   
5             3            9318                  2078    NaN         NaT   
6             2            8644                  2034    NaN         NaT   
7             1            8597                  2337    NaN         NaT   
8             1            9069                  2820    NaN         NaT   
9             1            9298                  2700    NaN         NaT   

   S_DEC_CESE  S_MTV_CESE S_SUB_MTV_CESE S_OBS_CESE  
0         NaN         NaN            NaN        NaN  
1         NaN         NaN            NaN        NaN  
2         NaN         NaN            NaN        NaN  
3         NaN         NaN            NaN        NaN  
4         NaN         NaN            NaN        NaN  
5         NaN         NaN            NaN        NaN  
6         NaN         NaN            NaN        NaN  
7         NaN         NaN            NaN        NaN  
8         NaN         NaN            NaN        NaN  
9         NaN         NaN            NaN        NaN  

[10 rows x 23 columns]

In [ ]:
base_final_merge['S_DEC_CESE'].value_counts()

S_DEC_CESE
COLABORADOR    562
BANCO          300
EXCLUIDO       214
Name: count, dtype: int64

In [ ]:
#AGREGAMOS VARIABLE BINARIA CON LA QUE DETERMINAREMOS SI EL EMPLEADO ES UN CESADO O NO
base_final_merge['S_DEC_CESE'] = base_final_merge['S_DEC_CESE'].fillna("")
base_final_merge['CESADO'] = base_final_merge['S_DEC_CESE'].apply(lambda x: 1 if x != "" else 0)
base_final_merge.head(10)

index  PERIODO  MATRICULA GENERO  EDAD GRUPO_GENERACION ESTADO_CIVIL  \
0  54435   202406      11060      F    49     GENERACION X  SOLTERO (A)   
1  53680   202406      12682      F    41      MILLENNIALS   CASADO (A)   
2  51612   202406      87728      M    37      MILLENNIALS   CASADO (A)   
3  51499   202406      62270      F    46     GENERACION X   CASADO (A)   
4  50509   202406      11423      F    43      MILLENNIALS  SOLTERO (A)   
5  58377   202406      82105      M    31      MILLENNIALS  SOLTERO (A)   
6  48968   202406      89420      F    38      MILLENNIALS   CASADO (A)   
7  58622   202406      97139      M    33      MILLENNIALS   CASADO (A)   
8  53539   202406      89458      F    39      MILLENNIALS   CASADO (A)   
9  56492   202406      89975      M    41      MILLENNIALS   CASADO (A)   

                                  TIPO_CONTRATO FECHA_CONTRATACION  \
0                                 INDETERMINADO         2007-03-26   
1  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES         2010-09-20   
2                        A PLAZO INDET - D.LEG.         2019-12-01   
3                                 INDETERMINADO         2012-02-15   
4                                 INDETERMINADO         2008-07-21   
5                                  DE SUPLENCIA         2018-07-25   
6                        A PLAZO INDET - D.LEG.         2021-02-01   
7                    A PLAZO INDET - D.LEG. 728         2024-06-01   
8                                 INDETERMINADO         2021-02-15   
9                        A PLAZO INDET - D.LEG.         2021-10-22   

  JORNADA_LABORAL  ... TOTAL_PUESTOS RANGO_SALARIAL  RANG_SALARIAL_INICIAL  \
0       FULL_TIME  ...             3           8947                   2563   
1       FULL_TIME  ...             3           8516                   2801   
2       FULL_TIME  ...             1           8701                   2877   
3       FULL_TIME  ...             4           9073                   2333   
4       FULL_TIME  ...             1           8756                   2978   
5       FULL_TIME  ...             3           9318                   2078   
6       FULL_TIME  ...             2           8644                   2034   
7       FULL_TIME  ...             1           8597                   2337   
8       FULL_TIME  ...             1           9069                   2820   
9       FULL_TIME  ...             1           9298                   2700   

  S_MAT  D_FEC_CESE  S_DEC_CESE  S_MTV_CESE  S_SUB_MTV_CESE S_OBS_CESE CESADO  
0   NaN         NaT                     NaN             NaN        NaN      0  
1   NaN         NaT                     NaN             NaN        NaN      0  
2   NaN         NaT                     NaN             NaN        NaN      0  
3   NaN         NaT                     NaN             NaN        NaN      0  
4   NaN         NaT                     NaN             NaN        NaN      0  
5   NaN         NaT                     NaN             NaN        NaN      0  
6   NaN         NaT                     NaN             NaN        NaN      0  
7   NaN         NaT                     NaN             NaN        NaN      0  
8   NaN         NaT                     NaN             NaN        NaN      0  
9   NaN         NaT                     NaN             NaN        NaN      0  

[10 rows x 24 columns]

In [ ]:
base_final_merge['CESADO'].value_counts()

CESADO
0    1173
1    1076
Name: count, dtype: int64

In [ ]:
import re
from collections import Counter

# Diccionario de palabras clave y sus respectivas categorías
keywords = {
    "Motivos Personales": ["personal", "familia", "personales"],
    "Oferta de Trabajo": ["oferta", "sueldo", "Mejor Oportunidad Laboral"],
    "Motivos de Salud": ["salud"],
    "Desempeño": ["desempeño"],
    "Cambio de Ubicación": ["cambio", "residencia", "provincia"],
    "Pase a Planilla": ["planilla", "Pase a planilla", "PASE A PLANILLA"],
    "Tercerización": ["tercerización"],
    "Perfil": ["perfil"],
    "Renuncia Voluntaria": ["voluntaria", "renuncia"],
    "Problemas Familiares": ["problemas familiares"],
    "Mutuo Acuerdo": ["mutuo acuerdo"],
    "Reestructuración": ["reestructuración"]
}

# Función para categorizar observaciones
def categorize_observation(text):
    if pd.isna(text):  # Dejar los NaN tal como están
        return np.NaN

    text = text.lower()  # Convertir a minúsculas para uniformidad
    for category, words in keywords.items():
        if any(re.search(rf'\b{word.lower()}\b', text) for word in words):
            return category
    return "Otros"  # Si no coincide con ninguna categoría

# Aplicar la función de categorización
base_final_merge['MOTIVO_CESE'] = base_final_merge['S_OBS_CESE'].apply(categorize_observation)



# Mostrar la distribución de categorías
print(base_final_merge['MOTIVO_CESE'].value_counts(dropna=False))

MOTIVO_CESE
NaN                    1240
Motivos Personales      321
Otros                   278
Renuncia Voluntaria     118
Perfil                   80
Oferta de Trabajo        54
Desempeño                39
Pase a Planilla          36
Mutuo Acuerdo            32
Cambio de Ubicación      27
Tercerización            17
Reestructuración          4
Motivos de Salud          3
Name: count, dtype: int64


In [ ]:
base_final.head(10)

index  PERIODO  MATRICULA GENERO  EDAD GRUPO_GENERACION ESTADO_CIVIL  \
1627  54435   202406      11060      F    49     GENERACION X  SOLTERO (A)   
1716  53680   202406      12682      F    41      MILLENNIALS   CASADO (A)   
2166  51612   202406      87728      M    37      MILLENNIALS   CASADO (A)   
2155  51499   202406      62270      F    46     GENERACION X   CASADO (A)   
1923  50509   202406      11423      F    43      MILLENNIALS  SOLTERO (A)   
1735  58377   202406      82105      M    31      MILLENNIALS  SOLTERO (A)   
1993  48968   202406      89420      F    38      MILLENNIALS   CASADO (A)   
1701  58622   202406      97139      M    33      MILLENNIALS   CASADO (A)   
2113  53539   202406      89458      F    39      MILLENNIALS   CASADO (A)   
1748  56492   202406      89975      M    41      MILLENNIALS   CASADO (A)   

                                     TIPO_CONTRATO FECHA_CONTRATACION  \
1627                                 INDETERMINADO         2007-03-26   
1716  PERS. CONVENIO DE PRACTICAS PREPROFESIONALES         2010-09-20   
2166                        A PLAZO INDET - D.LEG.         2019-12-01   
2155                                 INDETERMINADO         2012-02-15   
1923                                 INDETERMINADO         2008-07-21   
1735                                  DE SUPLENCIA         2018-07-25   
1993                        A PLAZO INDET - D.LEG.         2021-02-01   
1701                    A PLAZO INDET - D.LEG. 728         2024-06-01   
2113                                 INDETERMINADO         2021-02-15   
1748                        A PLAZO INDET - D.LEG.         2021-10-22   

     JORNADA_LABORAL                                             PUESTO  \
1627       FULL_TIME                             TRIBE TECHNICAL LEADER   
1716       FULL_TIME         GERENTE DE SERVICIO DE CANALES DE COBRANZA   
2166       FULL_TIME          GERENTE DE SERVICIO DE CONTROL FINANCIERO   
2155       FULL_TIME  GERENTE DE SERVICIO DE SOSTENIBILIDAD E INCLUS...   
1923       FULL_TIME        GERENTE DE SERVICIO DE RIESGOS DE TESORERIA   
1735       FULL_TIME  PRODUCT OWNER DE ESTRATEGIA MORA TARDIA DE REC...   
1993       FULL_TIME               PRODUCT OWNER DE PLANEAMIENTO Y DATA   
1701       FULL_TIME  GERENTE DE SERVICIO DE GESTION DE LA DEMANDA D...   
2113       FULL_TIME                               PRODUCT OWNER DE CRM   
1748       FULL_TIME                             TRIBE TECHNICAL LEADER   

     FECHA_INICIO_PUESTO  TIEMPO_PUESTO  \
1627          2023-09-15              9   
1716          2024-04-19              2   
2166          2019-12-01             54   
2155          2023-07-04             11   
1923          2016-07-11             95   
1735          2024-03-07              3   
1993          2022-04-04             26   
1701          2024-06-01              0   
2113          2021-04-22             38   
1748          2021-10-22             32   

                                               DIVISION  TOTAL_PUESTOS  \
1627                                         TECNOLOGIA              3   
1716                                            RIESGOS              3   
2166                               DIVISION DE FINANZAS              1   
2155  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...              4   
1923                                            RIESGOS              1   
1735                                            RIESGOS              3   
1993                        TRIBU SEGMENTOS Y PRODUCTOS              2   
1701                                         TECNOLOGIA              1   
2113  G.G ADJ. DE INTERNACIONALIZACION E INCLUSION F...              1   
1748                                         TECNOLOGIA              1   

      RANGO_SALARIAL  RANG_SALARIAL_INICIAL  
1627            8947                   2563  
1716            8516                   2801  
2166            8701                   2877  
2155            9073                   2333 

### ** RENTABILIDAD DE EMPLEADOS **
Agregamos columna extra de rentabilidad de los empleados

In [ ]:
# Calculamos los años en la empresa
base_final_merge['FECHA_CONTRATACION'] = pd.to_datetime(base_final_merge['FECHA_CONTRATACION'])
base_final_merge['D_FEC_CESE'] = pd.to_datetime(base_final_merge['D_FEC_CESE'])
base_final_merge['TIEMPO_EMPRESA'] = base_final_merge.apply(lambda row: (row['D_FEC_CESE'] - row['FECHA_CONTRATACION']).days / 365
                                    if pd.notnull(row['D_FEC_CESE'])
                                    else (pd.Timestamp.today() - row['FECHA_CONTRATACION']).days / 365, axis=1)

# Costo inicial por adquisición
costo_adquisicion = 3317

# Calculamos el salario anual de cada empleado
base_final_merge['SALARIO_ANUAL'] = base_final_merge['RANGO_SALARIAL'] * 12

# Costo total para empleados activos
base_final_merge['COSTO_TOTAL'] = base_final_merge['RANGO_SALARIAL'] * base_final_merge['TIEMPO_EMPRESA'] + costo_adquisicion

# Para los empleados cesados, agregamos el costo de reemplazo
base_final_merge['COSTO_REEMPLAZO'] = base_final_merge['CESADO'].apply(lambda x: 3317 if x == 1 else 0)

# Costo total acumulado (incluye el costo de reemplazo si el empleado fue cesado)
base_final_merge['COSTO_TOTAL_ACUMULADO'] = base_final_merge['COSTO_TOTAL'] + base_final_merge['COSTO_REEMPLAZO']

# Rentabilidad (para todos, activos y cesados)
base_final_merge['RENTABILIDAD'] = (base_final_merge['RANGO_SALARIAL']-base_final_merge['RANG_SALARIAL_INICIAL'])*base_final_merge['TIEMPO_EMPRESA'] / base_final_merge['COSTO_TOTAL_ACUMULADO'] * 100

print(base_final_merge[['MATRICULA', 'RENTABILIDAD','CESADO']])

      MATRICULA  RENTABILIDAD  CESADO
0         11060     69.874730       0
1         12682     65.295885       0
2         87728     62.033273       0
3         62270     72.195230       0
4         11423     64.480689       0
...         ...           ...     ...
2244      64440     31.755444       0
2245      75179     21.065191       0
2246      93227     10.099983       0
2247      87444     14.295812       0
2248      93803      9.526276       0

[2249 rows x 3 columns]


In [ ]:
print(max(base_final_merge['RENTABILIDAD']))
print(min(base_final_merge['RENTABILIDAD']))

82.53801867747586
-99.79191749545558


In [ ]:
#ELIMINAMOS LAS VARIABLES QUE NOS AYUDARON A OBTENER LA RENTABILIDAD
base_final_merge = base_final_merge.drop(columns=['COSTO_TOTAL','COSTO_REEMPLAZO','COSTO_TOTAL_ACUMULADO'])
base_final_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2249 entries, 0 to 2248
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   index                  2249 non-null   int64         
 1   PERIODO                2249 non-null   int64         
 2   MATRICULA              2249 non-null   int64         
 3   GENERO                 2249 non-null   object        
 4   EDAD                   2249 non-null   int64         
 5   GRUPO_GENERACION       2249 non-null   object        
 6   ESTADO_CIVIL           2249 non-null   object        
 7   TIPO_CONTRATO          2249 non-null   object        
 8   FECHA_CONTRATACION     2249 non-null   datetime64[ns]
 9   JORNADA_LABORAL        2249 non-null   object        
 10  PUESTO                 2249 non-null   object        
 11  FECHA_INICIO_PUESTO    2249 non-null   datetime64[ns]
 12  TIEMPO_PUESTO          2249 non-null   int64         
 13  DIV

In [ ]:
#GUARDAMOS LA BASE FINAL:
base_final_merge.to_csv('base_final2.csv', index=False)

In [ ]:
len(base_final['DIVISION'])

2249

#BASE FINAL PARA SU DESCARGA:
https://base-project-g14-iebs.s3.amazonaws.com/base_final2.csv

#BASE PRIMERIZA UNIFICADA:
https://base-project-g14-iebs.s3.amazonaws.com/base_unificada.csv